### DISCUSSION TITLE
selective loading and smart loading for large projects

### DISCUSSION THREAD

**Discussion by nullhook** [clangd/clangd#1632] [Ideas]
Currently, clangd loads all indexes stored on disk into memory. However, for large projects like Chromium, this can be excessive. 

<img width="658" alt="image" src="https://github.com/clangd/clangd/assets/8665427/e28a8ff8-62c6-4d9c-9b77-3bf574e27890">

Ideally, we want to address this issue by indexing all files on-disk and loading only the selected ones into memory. Another approach is to index all files on-disk and let clangd determine which ones should be loaded in-memory. This can be referred to as selective loading and smart loading, respectively. 

Open to ideas and thoughts here.

**Comment by HighCommander4** [comment]
"Selective loading" can be accomplished today, without code changes to clangd, by choosing where `compile_commands.json` files (henceforth, CDBs) are placed and what they contain. (But there are some caveats, see below.)

Note that:
 * Clangd determines which files to index based on which files are listed in the CDB, and the index is stored next to the CDB (in `.cache/clangd/index`).
 * Loading an index (as well as updating it, or indexing it for the first time) is triggered when the CDB is loaded, which happens when a source file in its directory is opened.

So you can accomplish "selective loading", at a directory granularity, as follows:
 * For subdirectories whose index you do not want to load by default (say, `<project>/a/foo` and `<project>/bar`), give them their own CDBs (`<project>/a/foo/compile_commands.json` and `<project>/bar/compile_commands.json`).
 * In the project's main CDB (`<project>/compile_commands.json`), list all source files **except** those in the above directories.

Note that this does not require the excluded subdirectories to be independently built; your build can generate a single CDB containing entries for the whole project, and you can post-process it to produce multiple filtered CDBs as described above by running a post-build step that does path-based filtering using something like [jq](https://stedolan.github.io/jq/).

---

This approach has some caveats:

 * `workspace/symbol` (e.g. vscode's "Go to Symbol in Workspace...") will only offer results from indexes that have been loaded
 * go-to-definition will not be able to navigate to definitions located in directories whose index has not been loaded

As such, I think it's suitable for an "I don't use these subdirectories very much" use case, but less so for an "I use the entire project but want parts of the index loaded on demand as I need them" use case.

It is possible to envision an alternative index architecture that involves:
 * Having one index for the entire project
 * Loading only a subset of it into memory and paging the rest in from disk on demand
 * Still having `workspace/symbol`, go-to-definition, etc. work across the entire project

Eclipse CDT has such an architecture; I'll describe it briefly:
 * The index consists of a set of lookup tables, and a set of records, persisted to disk separately
 * The lookup tables contain mappings from a key (e.g. a symbol name string) to an **offset** of a record
 * The lookup tables are loaded into memory entirely (but their total memory footprint is relatively small compared to the entire index)
 * The records themselves are "paged in" from disk on demand, when the result of a lookup in a lookup table indicates that they're needed

With this architecture, e.g. `workspace/symbol` works across the entire project because it involves looking up the search query string in the lookup table (which is in memory), and then loading just the target records from disk on demand.

**Comment by nullhook** [comment]
Thanks for the detailed writeup.

1. What does the current .idx file consist of?
2. What happens when the client opens a source/header file in the editor? Is the translation unit (TU) of this file parsed and examined to generate the resulting AST? If so, how long is this AST kept in memory? Is there a specific destroy policy for managing its lifetime?

**Reply by HighCommander4** [reply]
>    1. What does the current .idx file consist of?

You can see the contents [here](https://searchfox.org/llvm/rev/60d9010aaf0f09c6ccc4ae130be10f360f93612b/clang-tools-extra/clangd/index/Serialization.h#43-51).

>    2. What happens when the client opens a source/header file in the editor? Is the translation unit (TU) of this file parsed and examined to generate the resulting AST?

The AST is built directly from the source; the index is not involved.

Opening a file may trigger loading a whole index (all files in a `.cache/clangd/index`), if it's the first source file in that subdirectory to be opened.

The index is used to power cross-TU features such as find-references, often in combination with the AST (for example, for code completion, completion proposals from the current translation unit come from the AST and proposals from other translation units come from the index).

> If so, how long is this AST kept in memory? Is there a specific destroy policy for managing its lifetime?

To a first approximation, as long as the file is open.

(There are some slight nuances , e.g. the AST is split into a preamble portion which covers the contents of the includes at the top, and a main-file portion which covers the rest of the file. The preambles are retained as long as the file is open, but the main-file portions are kept in a small cache and may be evicted by merely switching away from an open file in the editor rather than closing it; however, this is not usually noticeable because the main-file portions are quick to rebuild compared to the preamble. There are also various proposals lying around to persist preambles **beyond** the file being closed, so that if you e.g. close and then reopen a file we don't have to rebuild the preamble on the reopen.)

**Comment by nullhook** [comment]
I'm curious if other maintainers here would like to share their opinions.

It appears that this proposal involves changes to the current structure of index files and other areas of clangd. If the team is interested, we can collaborate on a spec. I have enough capacity to implement this.

**Reply by HighCommander4** [reply]
To clarify, "this proposal" refers to the "alternative index architecture" I described in [this comment](https://github.com/clangd/clangd/discussions/1632#discussioncomment-5880954), with the goal of reducing clangd's memory usage by not having to load the entire index into memory at once (while preserving existing functionality).

Tagging @hokein @kadircet @sam-mccall for thoughts.
